NBTSerializable

Protonull opened the pull request:
Created a method of serializing and deserializing to and from NBTCompounds that mirrors the ConfigurationSerializable interface, but serializes into an NBTCompound rather than a Map.

Protonull committed:
Remove these system properties from surefire, they're not really needed

Protonull committed:
Creates an NBTSerializable interface and a manager

Protonull committed:
Unit tests for NBTSerializable

Protonull committed:
Bump version - 1.7.8

Protonull committed:
Merge pull request #3 from CivClassic/civ14

Civ14

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerializable.java:
```diff
@@ -0,0 +1,33 @@
+	 *
+	 * @throws NBTSerializationException This is thrown if the implementation has a fatal error serializing.
+	 * */
+	void serialize(NBTCompound nbt) throws NBTSerializationException;
```
If you assume the used NBTCompound to be empty, why not just create a new one in this method and return it?

Protonull replied:
> You can generally assume that the nbt compound is new and therefore empty

Working with a given NBTCompound rather than creating one's own allows classes to serialise themselves onto already populated compounds if the use case so warrants it. Whereas the method always creating its own compound and returning it assumes that you always, *always* want one compound to match one instance, which assumes that NBTSerializable will always be used just for NBTSerialization, which is not necessarily true.

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerializable.java:
```diff
@@ -0,0 +1,33 @@
+	void serialize(NBTCompound nbt) throws NBTSerializationException;
+
+	/**
+	 * Serialises a class into an NBTCompound.
```
Wrong docu

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java:
```diff
@@ -0,0 +1,160 @@
+public final class NBTSerialization {
+
+	private static final String NBT_CLASS_PATH_KEY = "==";
+    private static final Map<String, Class<? extends NBTSerializable>> registeredClasses = new HashMap<>();
```
weird indentation here

Protonull replied:
Weird, might need to re-apply the editorconfig to this file

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java:
```diff
@@ -0,0 +1,160 @@
+            throw new IllegalArgumentException("Cannot register NBTSerializable: " +
+                    "the given class is already registered.");
+        }
+        System.out.println("NBTSerializable[" + clazz.getName() + "] registered.");
```
Always use a logger over System.out

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java:
```diff
@@ -0,0 +1,160 @@
+                    "the given class is null.");
+        }
+        boolean hasValidConstructor = false;
+		for (Constructor<?> constructor : clazz.getDeclaredConstructors()) {
```
probably just format the whole file

Protonull replied:
Yeah, I'm really not sure what's going on here... this file is new and the only copy pastes were from within the file itself, which is how some of the wrong documentation came about... weird

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java:
```diff
@@ -0,0 +1,160 @@
+            String errorMessage = "Could not register the alias for NBTSerializable [" + clazz.getName() + "] as: ";
+            for (String alias : aliases) {
+                if (Strings.isNullOrEmpty(alias)) {
+                    System.err.println(errorMessage + "the alias is null or empty.");
```
same thing here and further down, use plugin.getLogger().severe()

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java:
```diff
@@ -0,0 +1,160 @@
+			serializable.serialize(nbt);
+		}
+		catch (Exception exception) {
+			exception.printStackTrace();
```
Same here and further down, use a logger

Maxopoly commented on src/test/java/vg/civcraft/mc/civmodcore/serialization/Tester.java:
```diff
@@ -0,0 +1,32 @@
+public class Tester {
+
+	@Test
+	public void testSerialization() {
```
nice

Protonull replied:
Thanks :) it does mean another dependency (junit) but I hope that's okay considering its purpose

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java:
```diff
@@ -0,0 +1,160 @@
+			instance = (T) clazz.newInstance();
+			instance.deserialize(nbt);
+		}
+		catch (Exception exception) {
```
Should do more explicit exception handling here and not just use the lazy "Exception"

Protonull replied:
Yeeaaah, I was kind of conflicted about this. Would a printing the exception suffice? I don't want the serialize and deserialize to be more lenient, especially since it'll likely be used within streams a lot, I know it already is in ItemExchange

Maxopoly replied:
Multi-try-Catch `(catch InvocationException|NoSuchMethodException|BlaBlaException e)` all the ones explicitly thrown by the reflection crap, log them and then throw a custom exception for the calling code. Anything going wrong here is pretty much guaranteed to be a bug in the calling code and should cause a RuntimeException.

Protonull replied:
I've kept the lazy catch all here and [here](https://github.com/CivClassic/CivModCore/pull/17#discussion_r366096900), because if anything throws an exception, then something has gone horribly wrong, so instead of just logging, a contextualised exception is thrown that contains the original exception.

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java:
```diff
@@ -0,0 +1,160 @@
+	 * casting the instance into the given generic type.
+	 * */
+	@SuppressWarnings("unchecked")
+    public static <T extends NBTSerializable> T deserialize(NBTCompound nbt) {
```
I think it would better to not make this method generic. You are using the generic as a convenience to cast the result to the desired type, which is not really necessary imo. If the given compound is of a different type than expected by the calling code, then that's a bug in the calling code and it should come up as such there when that code attempts to cast the returned object instead of having CMC swallow that error and return an ambigious `null`. 

So, I'd just return a NBTSerializable here.

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java:
```diff
@@ -0,0 +1,160 @@
+            return null;
+        }
+        String path = nbt.getString(NBT_CLASS_PATH_KEY);
+        nbt.remove(NBT_CLASS_PATH_KEY);
```
Neither do I think its a good decision design to have a deserialize method destroy the binary object in the process, nor do I think it should do so without documenting that.

Protonull replied:
The serialize method adds the class path, and the deserialize method removes it. The idea is to separate the class from the classpath process entirely, that the class can and should only know about the data itself.

Maxopoly replied:
Yes, but the problem there is that a use case might want to deserialize the same binary data multiple times, which wouldn't work with this.

Protonull replied:
That's... a fair point

Protonull replied:
After thinking about it for a while, I've decided to keep the `.remove()`, as I still think that the class being deserialised into should be ignorant of the deserialisation process itself. But you raised a good point that this should not destructive, especially if you wish to deserialise the same data multiple times, and so this is resolved by giving the class a copy of the data, as you don't want deserialisers to potentially corrupt the data for proceeding deserialisations.

Maxopoly reviewed (changes requested):
Overall good architectural decisions, excellent documentation and clean code. Only got some minor complaints here and there.

Protonull committed:
Fixed spelling

Protonull committed:
Merge branch 'civ14' of https://github.com/Protonull/CivModCore into civ14

Protonull committed:
Applied .editorconfig to the following files

Protonull committed:
Removed generic

Protonull committed:
Changed over to a logger implementation

Protonull committed:
Added slf4j-simple as a dependency just for junit testing

The implementation used by Bukkit is not initialised during the testing phase, so one must be provided lest there be errors in the console about a lack of a binding

Protonull committed:
Cleanup

Protonull committed:
Deserialisation suggestion

Protonull committed:
Oopsie, missed this

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/api/NBTCompound.java:
```diff
@@ -87,6 +87,15 @@ public void clear() {
+	 *
+	 * @return Returns a duplicated version of this NBT compound.
+	 * */
+	public NBTCompound duplicate() {
```
Java convention would be to implement https://docs.oracle.com/javase/7/docs/api/java/lang/Cloneable.html for this functionality

Protonull replied:
Done

Protonull committed:
Changed duplicate to clone

Protonull committed:
NBTCompound must implement Cloneable
